pkg/experiment/symbolizer/symbolizer.go

```diff
@@ -0,0 +1,844 @@
+	req := Request{
+		BuildID:    buildID,
+		BinaryName: binaryName,
+		Locations:  make([]*Location, len(locs)),
+	}
+
+	// Prepare locations for symbolization
+	for i, loc := range locs {
+		req.Locations[i] = &Location{
+			Address: loc.loc.Address,
+			Mapping: &pprof.Mapping{
+				Start:   mapping.MemoryStart,
+				Limit:   mapping.MemoryLimit,
+				Offset:  mapping.FileOffset,
+				BuildID: buildID,
+			},
+		}
+	}
+
+	return req
+}
+
+// updateProfileWithSymbols updates the profile with symbolization results
+func (s *ProfileSymbolizer) updateProfileWithSymbols(profile *googlev1.Profile, mapping *googlev1.Mapping, locs []locToSymbolize, symLocs []*Location) {
+	for i, symLoc := range symLocs {
+		if len(symLoc.Lines) > 0 {
+			locIdx := locs[i].idx
+			profile.Location[locIdx].Line = make([]*googlev1.Line, len(symLoc.Lines))
+
+			for j, line := range symLoc.Lines {
+				// Create or find function name in string table
+				nameIdx := s.findOrAddString(profile, line.FunctionName)
+				filenameIdx := s.findOrAddString(profile, line.FilePath)
+
+				// Create or find function
+				funcId := s.findOrAddFunction(profile, nameIdx, filenameIdx, line.LineNumber)
+
+				profile.Location[locIdx].Line[j] = &googlev1.Line{
+					FunctionId: funcId,
+					Line:       int64(line.LineNumber),
+				}
+			}
+		}
+	}
+
+	mapping.HasFunctions = true
+	mapping.HasFilenames = true
+	mapping.HasLineNumbers = true
+}
+
+// findOrAddString finds a string in the string table or adds it if not found
+func (s *ProfileSymbolizer) findOrAddString(profile *googlev1.Profile, str string) int64 {
+	for i, s := range profile.StringTable {
+		if s == str {
+			return int64(i)
+		}
+	}
+
+	idx := int64(len(profile.StringTable))
+	profile.StringTable = append(profile.StringTable, str)
+	return idx
+}
+
+// findOrAddFunction finds a function in the function table or adds it if not found
+func (s *ProfileSymbolizer) findOrAddFunction(profile *googlev1.Profile, nameIdx, filenameIdx int64, lineNumber uint64) uint64 {
+	maxFuncId := uint64(0)
+	for _, f := range profile.Function {
+		maxFuncId = max(maxFuncId, f.Id)
+		if f.Name == nameIdx && f.Filename == filenameIdx {
+			return f.Id
+		}
+	}
+
+	funcId := maxFuncId + 1
+	profile.Function = append(profile.Function, &googlev1.Function{
+		Id:        funcId,
+		Name:      nameIdx,
+		Filename:  filenameIdx,
+		StartLine: int64(lineNumber),
+	})
+
+	return funcId
+}
+
+func (s *ProfileSymbolizer) Symbolize(ctx context.Context, req *Request) error {
+	start := time.Now()
+	status := StatusSuccess
+	defer func() {
+		s.metrics.profileSymbolization.WithLabelValues(status).Observe(time.Since(start).Seconds())
+	}()
+
+	if s.checkSymbolCache(req) {
+		return nil
+	}
+
+	if s.checkLidiaTableCache(ctx, req) {
+		return nil
+	}
+
+	if s.checkObjectStoreCache(ctx, req) {
+		return nil
+	}
+
+	return s.fetchAndCacheFromDebuginfod(ctx, req, &status)
+}
```

Comment:
A comment explaining the sequence and cache layers would be helpful.

Reply:
We simplifed a lot.